### ISSUE TITLE
CAST to PostgreSQL enum type lacks type qualification

### ISSUE THREAD

**Issue by tbroyer** [jOOQ/jOOQ#10277]
### Expected behavior and actual behavior:

Using `DSL.cast()` with some database enum type, the generated SQL doesn't qualify the enum type, leading to errors (unless the schema is in the search_path; this is how we uncovered this bug actually: our app works OK for us developers and on our demo servers, but failed when deployed on our client's servers, because they use a different database user, whose name doesn't match that of the database schema).

jOOQ will generate `cast(?::"the_schema"."the_enum" as the_enum)` (or `cast(?::"the_schema"."the_enum"[] as the_enum[])` for an array, in our actual case), which will trigger an error `ERROR: type "the_enum" does not exist(..)`.
Note how the type in the cast is not qualified, whereas it's correctly qualified in typing the parameter.

Of course, the cast here is redundant (I'm almost certain it was necessary in an older version of jOOQ though, but at least we have an easy fix for our app), but interestingly we also have a domain type (`CREATE DOMAIN … AS text`) that in turn **requires** us to use a cast when used as an array, and in this case jOOQ correctly uses the qualified domain type: `cast(?::varchar[] as "the_schema"."the_domain"[])`.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

* have an enum type in your schema (`CREATE TYPE … AS ENUM …`) and a table using it for one of its columns
* use a `DSL.cast(…, MY_TABLE.MY_FIELD)` where the field is of the enum type
* have a search_path that doesn't include your schema

MCVE at https://github.com/atolcd-contrib/jOOQ-mcve/tree/issue-10277

### Versions:

- jOOQ: 3.12.4 and 3.13.2
- Java: OpenJDK 11
- Database (include vendor): Postgresql (reproduced on 11 and 12)
- OS: Linux (Arch Linux, and Docker's `openjdk:11`)
- JDBC Driver (include name if inofficial driver): `org.postgresql:postgresql` versions 42.2.9 and 42.2.14

**Comment by lukaseder** [comment]
Thanks a lot for your report. I will look into this soon. This one seems related: https://github.com/jOOQ/jOOQ/issues/10455

**Comment by lukaseder** [comment]
Finally looking into this now. We're not taking into account the rendering settings when rendering cast type names (or DDL type names for that matter) at all. For example, the name case, schema mapping, etc. don't work either: https://github.com/jOOQ/jOOQ/issues/9347

**Comment by lukaseder** [comment]
A currently failing integration test for this issue here will be:

```java
@Test
public void testPostgresCastEnumType() throws Exception {
    bindAndInline(ctx -> {
        try {
            ctx.setSchema(PUBLIC).execute();
            assertEquals(UBookStatus.ON_STOCK,
                ctx.fetchValue(select(cast(val("ON STOCK"), T_BOOK.STATUS.getDataType()))));

            ctx.setSchema(MULTI_SCHEMA).execute();
            assertEquals(UBookStatus.ON_STOCK,
                ctx.fetchValue(select(cast(val("ON STOCK"), T_BOOK.STATUS.getDataType()))));
        }
        finally {
            ctx.setSchema(PUBLIC).execute();
        }
    });
}
```

**Comment by lukaseder** [comment]
The bug can be fixed with existing API. But a more thorough fix would depend on letting `EnumType` extend `Qualified`: https://github.com/jOOQ/jOOQ/issues/10476

**Comment by lukaseder** [comment]
Huh, the `AbstractDataType::asEnumDataType` method can't read the `EnumType::getSchema`, currently, as our integration tests hint at a static initialisation problem:

```
Thread [main] (Suspended)	
	BuiltInDataType<T>(AbstractDataType<T>).asEnumDataType(Class<E>) line: 622	
	FEnumTable_6824.<init>(Name, Table<FEnumTable_6824Record>) line: 49	
	FEnumTable_6824.<init>() line: 75	
	FEnumTable_6824.<clinit>() line: 32	
	Public.<init>() line: 530	
	Public.<clinit>() line: 197	
	UCountry.getSchema() line: 38	
	BuiltInDataType<T>(AbstractDataType<T>).asEnumDataType(Class<E>) line: 621	
	FEnum_6502.<init>(Name, Table<FEnum_6502Record>, Field<?>[]) line: 45	
	FEnum_6502.<init>(Name, Table<FEnum_6502Record>) line: 48	
	FEnum_6502.<init>() line: 74	
	FEnum_6502.<clinit>() line: 32	
	Tables.<clinit>() line: 473	
	PostgresTest.TBook() line: 402	
	PostgresTest(jOOQAbstractTest<A,B,S,B2S,BS>).setUp() line: 665	
	NativeMethodAccessorImpl.invoke0(Method, Object, Object[]) line: not available [native method]	
	NativeMethodAccessorImpl.invoke(Object, Object[]) line: 77	
	DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: 43	
	Method.invoke(Object, Object...) line: 568	
	FrameworkMethod$1.runReflectiveCall() line: 59	
	FrameworkMethod$1(ReflectiveCallable).run() line: 12	
	FrameworkMethod.invokeExplosively(Object, Object...) line: 56	
	RunBefores.invokeMethod(FrameworkMethod) line: 33	
	RunBefores.evaluate() line: 24	
	RunAfters.evaluate() line: 27	
	TestWatcher$1.evaluate() line: 61	
	TestWatcher$1.evaluate() line: 61	
	ParentRunner$3.evaluate() line: 306	
	BlockJUnit4ClassRunner$1.evaluate() line: 100	
	JUnit4(ParentRunner<T>).runLeaf(Statement, Description, RunNotifier) line: 366	
	JUnit4(BlockJUnit4ClassRunner).runChild(FrameworkMethod, RunNotifier) line: 103	
	JUnit4(BlockJUnit4ClassRunner).runChild(Object, RunNotifier) line: 63	
	ParentRunner$4.run() line: 331	
	ParentRunner$1.schedule(Runnable) line: 79	
	JUnit4(ParentRunner<T>).runChildren(RunNotifier) line: 329	
	ParentRunner<T>.access$100(ParentRunner, RunNotifier) line: 66	
	ParentRunner$2.evaluate() line: 293	
	RunBefores.evaluate() line: 26	
	RunAfters.evaluate() line: 27	
	ParentRunner$3.evaluate() line: 306	
	JUnit4(ParentRunner<T>).run(RunNotifier) line: 413	
	JUnit4TestReference.run(TestExecution) line: 93	
	TestExecution.run(ITestReference[]) line: 40	
	RemoteTestRunner.runTests(String[], String, TestExecution) line: 529	
	RemoteTestRunner.runTests(TestExecution) line: 756	
	RemoteTestRunner.run() line: 452	
	RemoteTestRunner.main(String[]) line: 210	
```

**Comment by lukaseder** [comment]
That cycle can be prevented by introducing a new `LazyName <: Name` type, which evaluates its `transient` delegate `Name` object lazily upon first need. That way, we can delay the call to `EnumType::getSchema` until after class initialisation of the schema, at least until we need to access `Name` components in the `DefaultDataType` constructor.

**Comment by lukaseder** [comment]
The test will fix casts for PostgreSQL and related dialects only. Casting to an enum type in MySQL won't work, as the enum type is synthetic. Might fix that in a separate bug.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13360), and 3.15.10 (https://github.com/jOOQ/jOOQ/issues/13361)
